server/src/test/java/org/opensearch/cluster/service/MasterServiceTests.java

```diff
@@ -691,6 +691,266 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         }
     }
 
+    public void testThrottlingForTaskSubmission() throws InterruptedException {
+        int throttlingLimit = randomIntBetween(1, 10);
+        int taskId = 1;
+        final CyclicBarrier barrier = new CyclicBarrier(2);
+        final CountDownLatch latch = new CountDownLatch(1);
+        final String taskName = "test";
+        class Task {
+            private final int id;
+
+            Task(int id) {
+                this.id = id;
+            }
+        }
+
+        class TaskExecutor implements ClusterStateTaskExecutor<Task> {
+            private AtomicInteger published = new AtomicInteger();
+
+            @Override
+            public ClusterTasksResult<Task> execute(ClusterState currentState, List<Task> tasks) throws Exception {
+                latch.countDown();
+                barrier.await();
+                return ClusterTasksResult.<Task>builder().successes(tasks).build(currentState);
+            }
+
+            @Override
+            public String getMasterThrottlingKey() {
+                return taskName;
+            }
+
+            @Override
+            public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
+                published.incrementAndGet();
+            }
+        }
+
+        MasterService masterService = createClusterManagerService(true);
+        masterService.masterTaskThrottler.updateThrottlingLimit(taskName, throttlingLimit);
+
+        final ClusterStateTaskListener listener = new ClusterStateTaskListener() {
+            @Override
+            public void onFailure(String source, Exception e) {}
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {}
+        };
+
+        TaskExecutor executor = new TaskExecutor();
+        // submit one task which will be execution, post that will submit throttlingLimit tasks.
+        try {
+            masterService.submitStateUpdateTask(
+                taskName,
+                new Task(taskId++),
+                ClusterStateTaskConfig.build(randomFrom(Priority.values())),
+                executor,
+                listener
+            );
+        } catch (Exception e) {
+            throw new AssertionError(e);
+        }
+        // wait till task enter in execution.
+        latch.await();
+
+        for (int i = 0; i < throttlingLimit; i++) {
+            try {
+                masterService.submitStateUpdateTask(
+                    taskName,
+                    new Task(taskId++),
+                    ClusterStateTaskConfig.build(randomFrom(Priority.values())),
+                    executor,
+                    listener
+                );
+            } catch (Exception e) {
+                throw new AssertionError(e);
+            }
+        }
+
+        // we have one task in execution and tasks in queue so next task should throttled.
```

Comment:
just for my understanding : how are the `throttlingLimit` tasks waiting in queue ? wouldn't `barrier` just finish the `Tasks` when two are ready ?

Reply:
Master executes all task in single thread only, Once we submit the first task, other tasks will go in queue only as first task's execution is not completed and waiting at `barrier` as we have initialized the `barrier` with value 2.